Cleanup of CMake files and network bridge option

tobsan opened the pull request:
- The setup script does not output anything if all is fine
- The Netlink code now calculates ipv4 broadcast addr correctly
- The network configuration gives the attached device the name eth0

Signed-off-by: Tobias Olausson tobias.olausson@pelagicore.com

Commit:
Some changes to network setup

* The setup script does not output anything if all is fine
* The Netlink code now calculates ipv4 broadcast addr correctly
* The network configuration gives the attached device the name eth0

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

DunderRoffe commented on libsoftwarecontainer/softwarecontainer_network.conf.in:
```diff
@@ -1,2 +1,3 @@
 lxc.network.type = veth
 lxc.network.link = lxcbr0
```
Maybe we should set the name "lxcbr0" via a variable that defaults to "lxcbr0"? This way it is configurable by the integrators.

tobsan replied:
Yes, I think that would be even better actually. I can do some CMake magic to make it happen.

tobsan replied:
This is done now.

DunderRoffe commented on scripts/setup_softwarecontainer.sh:
```diff
@@ -17,17 +17,12 @@
+if ! [ -n "$BRIDGE_AVAILABLE" ]; then
     echo "$BRIDGE was NOT FOUND, attempting to add..."
     $BRCTL_CMD addbr $BRIDGE
     $BRCTL_CMD setfd $BRIDGE 0
```
Should not this script been removed by your work with netlink ?

tobsan replied:
Actually no, bridge setup and iptables stuff is not supported by the Netlink class, since it was not needed for removing busybox from the network gateway. It would be nice to add though (sometime).

thenor reviewed (approved):
Looks good to me.

kursatkobya commented on libsoftwarecontainer/src/netlink.cpp:
```diff
@@ -217,7 +217,7 @@ ReturnCode Netlink::linkUp(const int ifaceIndex)
 }
 
-ReturnCode Netlink::setIP(const int ifaceIndex, const in_addr ip, const int netmask)
+ReturnCode Netlink::setIP(const int ifaceIndex, const in_addr ip, const unsigned int netmask)
```
Would it be better to use ` const unsigned char` for `netmask` here ?

Commit:
Make network bridge configurable at compile time

SC will make LXC use a network bridge, which was previously hard coded
to lxcbr0. This commit keeps that as the default, but one has the ability
to change it at compile time.

Also restructure the CMake top-level file a bit.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

kursatkobya reviewed (approved):
I got one weak tentative idea about netmask argument, but it looks good to me.

Commit:
netlink: use unsigned char

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

Commit:
Moves some cmake macros into their own modules

Also tries to simplify the structure of our cmake files

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

Commit:
Makes creation of network bridge an option

With this, SoftwareContainer will not create a network bridge at runtime
if one is missing. Instead, it will print an error message and exit. Creation
of network bridge is done through a compile-time option given to cmake.

A consequence of this check is that the rather ugly assert in the workspace
code is removed, so it is up to the creator of the workspace object to call
checkWorkspace. Furthermore, it is the responsibility of the creator of
the agent object to make it check for workspace (see main.cpp) before
triggering any preloading of containers.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

breakreturn commented on CMakeLists.txt:
```diff
@@ -49,26 +55,57 @@ add_definitions(${DBusCpp_CFLAGS_OTHER})
+# All available options goes here
+#
+
+# First, options on enabling / disabling gateways at compile-time
```
I'm not a fan of the "first", "then", style. I've seen us do it before and to me it doesn't bring much more than overhead if things are moved around at some point.

tobsan replied:
Good point, will update.

breakreturn commented on agent/src/main.cpp:
```diff
@@ -104,6 +104,13 @@ int main(int argc, char **argv)
     }
 
     SoftwareContainerAgent agent(mainContext, preloadCount, shutdownContainers, timeout);
+    if (!agent.checkWorkspace()) {
```
The agent has a reference to a workspace and is also responsible for providing it to containers. With this check, the interface of the workspace is exposed to the user of the agent, which I don't think is good. I suggest we move checks for the workspace to the agent and have it self be responsible for checking its own pre-requisites. I.e the user of the agent really wants to do something else than worry about status of members of the agent, the check can be done when that something else is done by the user instead.

breakreturn commented on doc/conf.py.in:
```diff
@@ -39,6 +39,17 @@ extensions = [
+
+for key in substs.keys():
+    rst_epilog += ".. |%s| replace:: %s\n" % (key, substs[key])
+
```
Could you explain the purpose of the two for loops in the comments?

breakreturn commented:
A description of the new configuration option seems to be missing from the README.

Commit:
fix doc generation

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

breakreturn commented:
I think the whole bridge thing should be documented in the user docs, e.g. in the integration chapter, and possibly other places where it might be relevant.

Commit:
Adds docs, reworks init phase of agent

The agent constructor will now throw if it fails, instead of
having the user call some init code, or having the class
assert(false) like before.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

Commit:
agent: adds documentation

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

Commit:
Fix install of dbus conf + fix error in agent

The agent pointer would go out of scope immediately when the try scope
ended. Obviously a bad thing.

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

Commit:
Jenkinsfile: install SC to /usr

Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>
